implementations/rust/ockam/ockam_api/src/signer/types.rs

```diff
@@ -0,0 +1,180 @@
+use crate::{CowBytes, CowStr};
+use core::fmt;
+use data_encoding::BASE32_DNSSEC;
+use minicbor::{Decode, Encode};
+use ockam_core::compat;
+use ockam_core::compat::borrow::Cow;
+
+#[cfg(feature = "tag")]
+use crate::TypeTag;
+
+#[derive(Debug, Decode, Encode)]
+#[rustfmt::skip]
+#[cbor(map)]
+pub struct Credential<'a> {
+    #[cfg(feature = "tag")]
+    #[n(0)] tag: TypeTag<3796735>,
+    #[b(1)] attributes: CowBytes<'a>,
+    #[b(2)] signature: Signature<'a>
+}
+
+impl<'a> Credential<'a> {
+    pub(super) fn new(attributes: CowBytes<'a>, signature: Signature<'a>) -> Self {
+        Credential {
+            #[cfg(feature = "tag")]
+            tag: TypeTag,
+            attributes,
+            signature,
+        }
+    }
+
+    pub fn attributes(&self) -> &[u8] {
+        &self.attributes
+    }
+
+    pub fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    pub fn to_owned<'b>(&self) -> Credential<'b> {
+        Credential {
+            #[cfg(feature = "tag")]
+            tag: TypeTag,
+            attributes: self.attributes.to_owned(),
+            signature: self.signature.to_owned(),
+        }
+    }
+}
+
+impl fmt::Display for Credential<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // The derived `Encode` impl does not error and writing to a vector does not either:
+        let bytes = minicbor::to_vec(self).expect("encoding a credential never fails");
+        f.write_str(&BASE32_DNSSEC.encode(&bytes))
+    }
+}
+
+impl TryFrom<&str> for Credential<'_> {
+    type Error = InvalidCredential;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        let b = BASE32_DNSSEC
+            .decode(value.as_bytes())
+            .map_err(|e| InvalidCredential(e.into()))?;
+        let c = minicbor::decode::<Credential>(&b).map_err(|e| InvalidCredential(e.into()))?;
+        Ok(c.to_owned())
+    }
+}
+
+#[derive(Debug, Decode, Encode)]
+#[rustfmt::skip]
+pub struct Signature<'a> {
+    #[b(0)] ident: IdentityId<'a>,
+    #[b(1)] signature: CowBytes<'a>
+}
+
+impl<'a> Signature<'a> {
+    pub fn new<S>(id: IdentityId<'a>, sig: S) -> Self
```

Comment:
Can this be `IdentityIdentifier`?

Reply:
Long ago, when I started this, I went with this `IdentityId` because `IdentityIdentifier` always allocates. Maybe we can revisit this.

Reply:
Yes, you can improve `IdentityIdentifier` struct if needed